avocado/utils/pci.py

```diff
@@ -0,0 +1,291 @@
+    """
+    cmd = "ls -l /dev/disk/by-path/ | grep %s | awk '{print $NF}'" % device
+    devices = process.system_output(cmd, shell=True, ignore_status=True)
+    devices_list = []
+    if devices:
+        for dev in devices.splitlines():
+            cmd = "realpath -s /dev/disk/by-path/%s" % dev
+            dev = process.system_output(cmd, shell=True, ignore_status=True)
+            devices_list.append(dev.strip('\n'))
+    return devices_list
+
+
+def get_interfaces_in_device(device):
+    """
+    Gets ethernet interface in a pci device.
+
+    :parm device: pci device.
+
+    :return: list of interfaces in a pci device.
+    """
+    cmd = "ls -l /sys/class/net/ | grep %s | awk '{print $NF}'" % device
+    devices = process.system_output(cmd, shell=True, ignore_status=True)
+    interfaces_list = []
+    if devices:
+        for dev in devices.splitlines():
+            interfaces_list.append(dev.split("/")[-1])
+    return interfaces_list
+
+
+def get_pci_fun_list(device):
+    """
+    Gets list of functions in the given pci device.
+    Example: in device 0000:03:00, functions are 0000:03:00.0 and 0000:03:00.1
+
+    :parm device: pci device.
+
+    :return: list of functions in a pci device.
+    """
+    cmd = "lspci -D | grep %s | awk {'print $1'}" % device
+    functions = process.system_output(cmd, shell=True, ignore_status=True)
+    if functions:
+        return functions.splitlines()
+
+
+def get_slot_from_sysfs(device):
+    """
+    Gets the pci slot of given device.
+
+    :parm device: pci device.
+
+    :return: slot of pci device from sysfs.
+    """
+    if not os.path.isdir('/sys/bus/pci/devices/%s' % device):
+        return
+    cmd = "cat /sys/bus/pci/devices/%s/devspec ; echo" % device
+    devspec = process.system_output(cmd, shell=True).strip('\n')
+    cmd = "cat /proc/device-tree/%s/ibm,loc-code ; echo" % devspec
+    slot = process.system_output(cmd, shell=True).strip('\0\n')
+    return re.match(r'((\w+)[\.])+(\w+)-(\w*\d+)-(\w*\d+)|Slot(\d+)',
+                    slot).group()
+
+
+def get_pci_id_from_sysfs(device):
+    """
+    Gets the pci id from sysfs of given device.
+
+    :parm device: pci device.
+
+    :return: pci id of a pci device from sysfs.
+    """
+    if not os.path.isdir('/sys/bus/pci/devices/%s' % device):
+        return
+    pci_id = ""
+    for params in ['vendor', 'device', 'subsystem_vendor', 'subsystem_device']:
+        cmd = "cat /sys/bus/pci/devices/%s/%s" % (device, params)
+        p_id = process.system_output(cmd, shell=True).strip('\n').strip('0x')
+        while len(p_id) < 4:
+            p_id = "0%s" % p_id
+        pci_id += "%s:" % p_id
+    if pci_id:
+        return pci_id[:-1]
+
+
+def get_slot_list():
+    """
+    Gets list of pci slots in the system.
+
+    :return: list of slots in the system.
+    """
+    slot_list = []
+    for device in get_pci_devices():
+        slot_list.append(get_slot_from_sysfs(device))
+    return list(set(slot_list))
```

Comment:
How about `return list(set(get_slot_from_sysfs(device) for device in get_pci_devices))`?